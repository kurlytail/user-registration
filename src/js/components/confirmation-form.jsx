import React, { Component } from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';

class RegistrationConfirmationForm extends Component {
    constructor(props) {
        super(props);
        this.state = { hash: props.hash, email: props.email };
    }

    handlePassword(event) {
        this.setState({ password: event.target.value });
    }

    handleConfirmPassword(event) {
        this.setState({ confirmPassword: event.target.value });
    }

    handleName(event) {
        this.setState({ name: event.target.value });
    }

    onSubmit(e) {
        e.preventDefault();
        this.props.actions.deleteRegistration({ ...this.state });
    }

    render() {
        let fieldErrors = _.get(this.props.state, 'error.errors');
        fieldErrors = fieldErrors ? fieldErrors : [];

        return (
            <form className="auth-form" method="POST" id="user-registration-form" encType="utf8">
                <div className="form-group">
                    <div className="form-label-group">
                        <label>email</label>
                        <input
                            type="email"
                            className="form-control"
                            id="user-registration-email"
                            readOnly
                            value={this.state.email}
                        />
                    </div>
                    <div className="form-label-group">
                        <label>Name</label>
                        <input
                            type="text"
                            className="form-control"
                            required
                            autoFocus
                            id="user-registration-name"
                            onChange={event => this.handleName(event)}
                            value={this.state.name}
                        />
                        {fieldErrors.map(
                            errorItem =>
                                errorItem.field === 'name' ? (
                                    <p className="text-red" key={errorItem.code}>
                                        {errorItem.defaultMessage}
                                    </p>
                                ) : (
                                    ''
                                )
                        )}
                    </div>
                    <div className="form-label-group">
                        <label>Password</label>
                        <input
                            type="password"
                            className="form-control"
                            required
                            id="user-registration-password"
                            onChange={event => this.handlePassword(event)}
                            value={this.state.password}
                        />
                        {fieldErrors.map(
                            errorItem =>
                                errorItem.field === 'password' ? (
                                    <p className="text-red" key={errorItem.code}>
                                        {errorItem.defaultMessage}
                                    </p>
                                ) : (
                                    ''
                                )
                        )}
                    </div>
                    <div className="form-label-group">
                        <label>Confirm Password</label>
                        <input
                            type="password"
                            className="form-control"
                            required
                            id="user-registration-confirmPassword"
                            onChange={event => this.handleConfirmPassword(event)}
                            value={this.state.confirmPassword}
                        />
                        {fieldErrors.map(
                            errorItem =>
                                errorItem.field === 'confirmPassword' ? (
                                    <p className="text-red" key={errorItem.code}>
                                        {errorItem.defaultMessage}
                                    </p>
                                ) : (
                                    ''
                                )
                        )}
                    </div>
                </div>
                <div className="form-group">
                    <button
                        className="btn btn-lg btn-primary btn-block"
                        type="submit"
                        id="signup-button"
                        onClick={e => this.onSubmit(e)}
                    >
                        Confirm Registration
                    </button>
                </div>
            </form>
        );
    }
}

RegistrationConfirmationForm.propTypes = {
    hash: PropTypes.string,
    email: PropTypes.string,
    state: PropTypes.any,
    actions: PropTypes.any
};

export default RegistrationConfirmationForm;
