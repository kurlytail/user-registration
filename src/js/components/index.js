import RegistrationForm from './registration-form';
import ConfirmationForm from './confirmation-form';

export { RegistrationForm, ConfirmationForm };
