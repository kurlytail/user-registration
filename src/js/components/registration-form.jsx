import React, { Component } from 'react';
import ReCAPTCHA from 'react-google-recaptcha';
import PropTypes from 'prop-types';
import _ from 'lodash';

class RegistrationForm extends Component {
    constructor(props) {
        super(props);
        this.state = {};
        this._reCaptchaRef = React.createRef();
    }

    handleEmail(event) {
        this.setState({ email: event.target.value });
    }

    onSubmit(e) {
        e.preventDefault();
        this.props.actions.createRegistration({ ...this.state, captchaKey: this.props.state.item.captchaKey });
    }

    render() {
        let fieldErrors = _.get(this.props.state, 'error.errors');
        fieldErrors = fieldErrors ? fieldErrors : [];

        return (
            <form className="auth-form" method="POST" id="user-registration-form" encType="utf8">
                <div className="form-group">
                    <div className="form-label-group">
                        <label>email</label>
                        <input
                            type="email"
                            className="form-control"
                            required
                            autoFocus
                            id="user-registration-email"
                            onChange={event => this.handleEmail(event)}
                        />
                        {fieldErrors.map(
                            errorItem =>
                                errorItem.field === 'email' ? (
                                    <p className="text-red" key={errorItem.code}>
                                        {errorItem.defaultMessage}
                                    </p>
                                ) : (
                                    ''
                                )
                        )}
                    </div>
                </div>
                <div className="form-group">
                    {_.get(this.props.state, 'item.captchaKey') && !_.get(this.props.state, 'isFetching') ? (
                        <ReCAPTCHA
                            id="user-registration-captcha"
                            sitekey={_.get(this.props.state, 'item.captchaKey')}
                            onChange={value => this.dataCallback(value)}
                            onExpired={() => this.dataExpiredCallback()}
                            onError={() => this.dataErrorCallback()}
                            ref={this._reCaptchaRef}
                        />
                    ) : (
                        <div />
                    )}
                    {fieldErrors.map(
                        errorItem =>
                            errorItem.field === 'reCaptchaResponse' ? (
                                <p className="text-red" key={errorItem.code}>
                                    {errorItem.defaultMessage}
                                </p>
                            ) : (
                                ''
                            )
                    )}
                </div>
                <div className="form-group">
                    <button
                        className="btn btn-lg btn-primary btn-block"
                        type="submit"
                        id="signup-button"
                        disabled={!this.state.reCaptchaResponse || !this.state.email}
                        onClick={e => this.onSubmit(e)}
                    >
                        Sign Up
                    </button>
                </div>
            </form>
        );
    }

    dataCallback(value) {
        this.setState({ reCaptchaResponse: value });
    }

    dataExpiredCallback() {
        this.setState({ reCaptchaResponse: null });
    }

    dataErrorCallback() {
        this.setState({ reCaptchaResponse: null });
    }
}

RegistrationForm.propTypes = {
    state: PropTypes.any,
    actions: PropTypes.any
};

export default RegistrationForm;
