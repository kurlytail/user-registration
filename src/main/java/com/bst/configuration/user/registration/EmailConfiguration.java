package com.bst.configuration.user.registration;

import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.mail.javamail.JavaMailSender;
import org.springframework.mail.javamail.JavaMailSenderImpl;

@Configuration
public class EmailConfiguration {
	
	@Bean(name="javaMailSender")
	public JavaMailSender createJavaMailSender() {
		return new JavaMailSenderImpl();
	}
}
