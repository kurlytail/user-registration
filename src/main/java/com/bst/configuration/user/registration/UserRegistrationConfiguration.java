package com.bst.configuration.user.registration;

import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
import org.springframework.context.annotation.ComponentScan;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.EnableAspectJAutoProxy;
import org.springframework.context.annotation.ImportResource;

@Configuration
@ComponentScan({"com.bst.user.registration.components", "com.bst.user.registration.configuration"})
@ImportResource({"classpath:Aspect.xml"})
@EnableAspectJAutoProxy
@EnableAutoConfiguration
public class UserRegistrationConfiguration {
}
