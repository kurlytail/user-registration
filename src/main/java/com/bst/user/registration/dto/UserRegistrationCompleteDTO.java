package com.bst.user.registration.dto;

import javax.validation.constraints.Email;
import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;

import com.bst.user.registration.constraints.ValidConfirmationHash;


@ValidConfirmationHash
public class UserRegistrationCompleteDTO {

	@Email
	@NotNull
	@NotEmpty
	private String email;
	
	@NotNull
	@NotEmpty
	private String name;

	@NotNull
	@NotEmpty
	private String password;
	
	@NotNull
	@NotEmpty
	private String confirmPassword;
	
	@NotNull
	@NotEmpty
	private String hash;

	public String getHash() {
		return hash;
	}

	public void setHash(String hash) {
		this.hash = hash;
	}

	public String getEmail() {
		return email;
	}

	public void setEmail(String email) {
		this.email = email;
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public String getPassword() {
		return password;
	}

	public void setPassword(String password) {
		this.password = password;
	}

	public String getConfirmPassword() {
		return confirmPassword;
	}

	public void setConfirmPassword(String confirmPassword) {
		this.confirmPassword = confirmPassword;
	}
}
