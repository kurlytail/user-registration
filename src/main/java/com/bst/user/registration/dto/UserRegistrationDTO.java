package com.bst.user.registration.dto;

import javax.validation.constraints.Email;
import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;

import com.bst.utility.constraints.ValidReCaptcha;

public class UserRegistrationDTO {
     
    @Email
    @NotNull
    @NotEmpty
    private String email;
    
    @ValidReCaptcha
    private String reReCaptchaResponse;

	public String getEmail() {
		return email;
	}

	public void setEmail(String email) {
		this.email = email;
	}

	public String getReCaptchaResponse() {
		return reReCaptchaResponse;
	}

	public void setReCaptchaResponse(String reReCaptchaResponse) {
		this.reReCaptchaResponse = reReCaptchaResponse;
	}

}
