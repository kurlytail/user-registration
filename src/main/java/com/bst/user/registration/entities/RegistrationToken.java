package com.bst.user.registration.entities;

import java.util.Date;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Transient;

import org.apache.commons.codec.digest.DigestUtils;

@Entity
public class RegistrationToken {

	@Id
	@GeneratedValue(strategy = GenerationType.AUTO)
	private Long id;
	private Date createdDate = new Date();
	@Column(unique = true)
	private String email;

	@Transient
	public String getHash() {
		return DigestUtils.sha512Hex(getEmail() + getCreatedDate() + getId().toString());
	}

	public RegistrationToken() {
		this.email = null;
	}

	public RegistrationToken(String email) {
		this.email = email;
	}

	public Long getId() {
		return id;
	}

	public Date getCreatedDate() {
		return createdDate;
	}

	public String getEmail() {
		return email;
	}
}
