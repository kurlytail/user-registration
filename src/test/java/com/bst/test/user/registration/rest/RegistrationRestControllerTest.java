package com.bst.test.user.registration.rest;

import java.util.Locale;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.ArgumentMatchers;
import org.mockito.Captor;
import org.mockito.Mockito;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.context.SpringBootTest.WebEnvironment;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.ResponseEntity;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.TestExecutionListeners;
import org.springframework.test.context.TestExecutionListeners.MergeMode;
import org.springframework.test.context.TestPropertySource;
import org.springframework.test.context.junit.jupiter.SpringExtension;

import com.bst.configuration.user.registration.UserRegistrationConfiguration;
import com.bst.test.user.registration.TestAppConfiguration;
import com.bst.test.user.registration.components.SecurityDisabler;
import com.bst.user.authentication.components.UserService;
import com.bst.user.registration.dto.UserRegistrationCompleteDTO;
import com.bst.user.registration.dto.UserRegistrationDTO;
import com.bst.user.registration.entities.RegistrationToken;
import com.bst.utility.configuration.CaptchaSettings;
import com.bst.utility.services.EmailService;
import com.bst.utility.services.ReCaptchaService;
import com.bst.utility.testlib.SnapshotListener;

@ExtendWith(SpringExtension.class)
@SpringBootTest(webEnvironment = WebEnvironment.RANDOM_PORT)
@TestExecutionListeners(listeners = SnapshotListener.class, mergeMode = MergeMode.MERGE_WITH_DEFAULTS)
@ContextConfiguration(classes = { UserRegistrationConfiguration.class, TestAppConfiguration.class,
		SecurityDisabler.class })
@TestPropertySource("classpath:registration-controller-test.properties")
public class RegistrationRestControllerTest {
	@MockBean
	private CaptchaSettings captchaSettings;

	@MockBean
	private EmailService emailService;

	@Captor
	ArgumentCaptor<RegistrationToken> hashCaptor;

	@Captor
	ArgumentCaptor<Locale> locale;

	@MockBean
	private ReCaptchaService reCaptchaService;

	@Captor
	ArgumentCaptor<String[]> recepients;

	private HttpHeaders requestHeaders;

	@Autowired
	private TestRestTemplate testRestTemplate;

	@MockBean
	private UserService userService;

	@BeforeEach
	public void setup() {
		Mockito.when(this.captchaSettings.getKey()).thenReturn("test captcha key");
	}

	@Test
	public void testRegistrationStates() throws Exception {
		final ResponseEntity<UserRegistrationDTO> getResponse = this.testRestTemplate.getForEntity("/registration",
				UserRegistrationDTO.class);
		SnapshotListener.expect(getResponse.getStatusCode()).toMatchSnapshot();
		SnapshotListener.expect(getResponse.getBody()).toMatchSnapshot();

		final UserRegistrationDTO dto = new UserRegistrationDTO();
		dto.setCaptchaKey("test captcha key");
		dto.setEmail("john@doe.com");
		dto.setReCaptchaResponse("test captcha response");

		final ResponseEntity<Object> postResponse = this.testRestTemplate.exchange("/registration", HttpMethod.POST,
				new HttpEntity<>(dto, this.requestHeaders), Object.class);
		SnapshotListener.expect(postResponse.getStatusCode()).toMatchSnapshot();
		SnapshotListener.expect(postResponse.getBody()).toMatchSnapshot();

		final UserRegistrationCompleteDTO completeDto = new UserRegistrationCompleteDTO();
		completeDto.setConfirmPassword("password");
		completeDto.setPassword("password");

		Mockito.verify(this.emailService).sendMessage(this.recepients.capture(), ArgumentMatchers.anyString(),
				ArgumentMatchers.anyString(), ArgumentMatchers.anyString(), ArgumentMatchers.anyString(),
				this.hashCaptor.capture(), this.locale.capture());
		completeDto.setHash(this.hashCaptor.getValue().getHash().toString());
		completeDto.setEmail("john@doe.com");
		completeDto.setName("John Doe");

		final ResponseEntity<Object> deleteResponse = this.testRestTemplate.exchange("/registration", HttpMethod.DELETE,
				new HttpEntity<>(completeDto, this.requestHeaders), Object.class);
		SnapshotListener.expect(deleteResponse.getStatusCode()).toMatchSnapshot();
		SnapshotListener.expect(deleteResponse.getBody()).toMatchSnapshot();
	}
}
