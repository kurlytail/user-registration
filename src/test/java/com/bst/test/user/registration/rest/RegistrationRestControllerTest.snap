{
  "testRegistrationStates" : [ "OK", {
    "captchaKey" : "test captcha key"
  }, "OK", { }, "OK", { } ]
}